dbms/src/Functions/FunctionsLogical.h

```diff
@@ -377,154 +504,470 @@ class FunctionAnyArityLogical : public IFunction
+        auto col_res = ColumnUInt8::create(rows, static_cast<UInt8>(0));
+        UInt8Container & vec_res = col_res->getData();
+        auto col_res_is_null = ColumnUInt8::create();
+        UInt8Container & vec_res_is_null = col_res_is_null->getData();
+        if (result_is_nullable)
+            vec_res_is_null.assign(rows, static_cast<UInt8>(0));
+
+        bool has_not_null_column;
+        // first handle not-null column
+        if (!not_null_uint8_columns.empty())
+        {
+            has_not_null_column = true;
+            if (not_null_uint8_columns.size() == 1)
+            {
+                const auto & col_data = not_null_uint8_columns[0]->getData();
+                // according to https://github.com/pingcap/tiflash/issues/5849
+                // need to cast the UInt8 column to bool explicitly
+                for (size_t i = 0; i < rows; ++i)
+                {
+                    vec_res[i] = static_cast<bool>(col_data[i]);
+                }
+            }
+            else
+            {
+                while (not_null_uint8_columns.size() > 1)
+                {
+                    // micro benchmark shows 4 << 6, 8 outperforms 6 slightly, and performance may decline when set to 10
+                    AssociativeOperationImpl<Impl, 8>::execute(not_null_uint8_columns, vec_res);
+                    not_null_uint8_columns.push_back(col_res.get());
+                }
+            }
+        }
+        // then handle nullable column
+        if (!nullable_uint8_columns.empty())
         {
-            /// With a large block size, combining 6 columns per pass is the fastest.
-            /// When small - more, is faster.
-            AssociativeOperationImpl<Impl, 10>::execute(uint8_in, vec_res);
-            uint8_in.push_back(col_res.get());
+            if (has_not_null_column)
+            {
+                // if there is not null column, then need to append the current result
+                // to uint8_not_null_column as the input column
+                nullable_uint8_columns.push_back(col_res.get());
+                null_maps.push_back(&vec_res_is_null);
+            }
+            if (nullable_uint8_columns.size() == 1)
+            {
+                // special case
+                assert(!has_not_null_column);
+                const auto & col_data = nullable_uint8_columns[0]->getData();
+                // according to https://github.com/pingcap/tiflash/issues/5849
+                // need to cast the UInt8 column to bool explicitly
+                for (size_t i = 0; i < rows; ++i)
+                {
+                    vec_res[i] = static_cast<bool>(col_data[i]);
+                }
+                vec_res_is_null.assign(*null_maps[0]);
+            }
+            else
+            {
+                while (nullable_uint8_columns.size() > 1)
+                {
+                    // micro benchmark shows 4 << 6, 8 outperforms 6 slightly, and performance may decline when set to 10
```

Comment:
Interesting results. Would you be willing to add some concrete numbers?

Reply:
Here is one of the benchmark results, block row size is 10000
```
Run on (72 X 3300 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x36)
  L1 Instruction 32 KiB (x36)
  L2 Unified 1024 KiB (x36)
  L3 Unified 25344 KiB (x2)
Load Average: 14.19, 14.67, 14.33
LogicalOpBench/optLogicalMultiParam_And6_not_null10/iterations:1000                     30615 ns        30481 ns         1000
LogicalOpBench/optLogicalMultiParam_And6_not_null8/iterations:1000                      30045 ns        29925 ns         1000
LogicalOpBench/optLogicalMultiParam_And6_not_null6/iterations:1000                      34859 ns        34705 ns         1000
LogicalOpBench/optLogicalMultiParam_And6_not_null4/iterations:1000                      48457 ns        48274 ns         1000
LogicalOpBench/optLogicalMultiParam_And7_not_null10/iterations:1000                     46847 ns        46678 ns         1000
LogicalOpBench/optLogicalMultiParam_And7_not_null8/iterations:1000                      49235 ns        49053 ns         1000
LogicalOpBench/optLogicalMultiParam_And7_not_null6/iterations:1000                      44633 ns        44456 ns         1000
LogicalOpBench/optLogicalMultiParam_And7_not_null4/iterations:1000                      72414 ns        72134 ns         1000
LogicalOpBench/optLogicalMultiParam_And8_not_null10/iterations:1000                     64318 ns        64091 ns         1000
LogicalOpBench/optLogicalMultiParam_And8_not_null8/iterations:1000                      61097 ns        60847 ns         1000
LogicalOpBench/optLogicalMultiParam_And8_not_null6/iterations:1000                      62362 ns        62137 ns         1000
LogicalOpBench/optLogicalMultiParam_And8_not_null4/iterations:1000                      84139 ns        83801 ns         1000
LogicalOpBench/optLogicalMultiParam_And9_not_null10/iterations:1000                     70640 ns        70362 ns         1000
LogicalOpBench/optLogicalMultiParam_And9_not_null8/iterations:1000                      77622 ns        77322 ns         1000
LogicalOpBench/optLogicalMultiParam_And9_not_null6/iterations:1000                      88216 ns        87870 ns         1000
LogicalOpBench/optLogicalMultiParam_And9_not_null4/iterations:1000                      98208 ns        97835 ns         1000
LogicalOpBench/optLogicalMultiParam_And10_not_null10/iterations:1000                    79753 ns        79401 ns         1000
LogicalOpBench/optLogicalMultiParam_And10_not_null8/iterations:1000                     97168 ns        96756 ns         1000
LogicalOpBench/optLogicalMultiParam_And10_not_null6/iterations:1000                    111726 ns       111300 ns         1000
LogicalOpBench/optLogicalMultiParam_And10_not_null4/iterations:1000                    114488 ns       114067 ns         1000
LogicalOpBench/optLogicalMultiParam_And11_not_null10/iterations:1000                    91365 ns        91036 ns         1000
LogicalOpBench/optLogicalMultiParam_And11_not_null8/iterations:1000                    109042 ns       108619 ns         1000
LogicalOpBench/optLogicalMultiParam_And11_not_null6/iterations:1000                    124121 ns       123599 ns         1000
LogicalOpBench/optLogicalMultiParam_And11_not_null4/iterations:1000                    127939 ns       127453 ns         1000
LogicalOpBench/optLogicalMultiParam_And6_nullable10/iterations:1000                     98956 ns        98600 ns         1000
LogicalOpBench/optLogicalMultiParam_And6_nullable8/iterations:1000                     101464 ns       101110 ns         1000
LogicalOpBench/optLogicalMultiParam_And6_nullable6/iterations:1000                     110970 ns       110589 ns         1000
LogicalOpBench/optLogicalMultiParam_And6_nullable4/iterations:1000                     112874 ns       112427 ns         1000
LogicalOpBench/optLogicalMultiParam_And7_nullable10/iterations:1000                    109919 ns       109538 ns         1000
LogicalOpBench/optLogicalMultiParam_And7_nullable8/iterations:1000                     109301 ns       108919 ns         1000
LogicalOpBench/optLogicalMultiParam_And7_nullable6/iterations:1000                     123537 ns       123085 ns         1000
LogicalOpBench/optLogicalMultiParam_And7_nullable4/iterations:1000                     135217 ns       134737 ns         1000
LogicalOpBench/optLogicalMultiParam_And8_nullable10/iterations:1000                    123250 ns       122759 ns         1000
LogicalOpBench/optLogicalMultiParam_And8_nullable8/iterations:1000                     136376 ns       135838 ns         1000
LogicalOpBench/optLogicalMultiParam_And8_nullable6/iterations:1000                     140589 ns       140092 ns         1000
LogicalOpBench/optLogicalMultiParam_And8_nullable4/iterations:1000                     148977 ns       148454 ns         1000
LogicalOpBench/optLogicalMultiParam_And9_nullable10/iterations:1000                    141426 ns       140945 ns         1000
LogicalOpBench/optLogicalMultiParam_And9_nullable8/iterations:1000                     148438 ns       147214 ns         1000
LogicalOpBench/optLogicalMultiParam_And9_nullable6/iterations:1000                     207442 ns       206634 ns         1000
LogicalOpBench/optLogicalMultiParam_And9_nullable4/iterations:1000                     173037 ns       172412 ns         1000
LogicalOpBench/optLogicalMultiParam_And10_nullable10/iterations:1000                   164605 ns       164018 ns         1000
LogicalOpBench/optLogicalMultiParam_And10_nullable8/iterations:1000                    168543 ns       167935 ns         1000
LogicalOpBench/optLogicalMultiParam_And10_nullable6/iterations:1000                    168199 ns       167608 ns         1000
LogicalOpBench/optLogicalMultiParam_And10_nullable4/iterations:1000                    194473 ns       193743 ns         1000
LogicalOpBench/optLogicalMultiParam_And11_nullable10/iterations:1000                   180895 ns       180236 ns         1000
LogicalOpBench/optLogicalMultiParam_And11_nullable8/iterations:1000                    178976 ns       178328 ns         1000
LogicalOpBench/optLogicalMultiParam_And11_nullable6/iterations:1000                    191045 ns       190356 ns         1000
LogicalOpBench/optLogicalMultiParam_And11_nullable4/iterations:1000                    204776 ns       204063 ns         1000
```
optLogicalMultiParam_And6_not_null10 means 6 not null input columns, and template argument is 10
optLogicalMultiParam_And11_nullable10 means 11 nullable input columns, and template argument is 10